test/e2e/kubevirt.go

```diff
@@ -0,0 +1,628 @@
+		forcePostCopyMigrationPolicy := &kvmigrationsv1alpha1.MigrationPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "force-post-copy",
+			},
+			Spec: kvmigrationsv1alpha1.MigrationPolicySpec{
+				AllowPostCopy:           pointer.Bool(true),
+				CompletionTimeoutPerGiB: pointer.Int64(1),
+				BandwidthPerMigration:   &bandwidthPerMigration,
+				Selectors: &kvmigrationsv1alpha1.Selectors{
+					VirtualMachineInstanceSelector: kvmigrationsv1alpha1.LabelSelector{
+						"test-live-migration": "post-copy",
+					},
+				},
+			},
+		}
+		if td.mode == kubevirtv1.MigrationPostCopy {
+			_, err = kvcli.MigrationPolicy().Create(context.TODO(), forcePostCopyMigrationPolicy, metav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			defer func() {
+				Expect(kvcli.MigrationPolicy().Delete(context.TODO(), forcePostCopyMigrationPolicy.Name, metav1.DeleteOptions{})).To(Succeed())
+			}()
+		}
+
+		By("Creating a test pod on both selected worker nodes")
+		podWorker1 = composeAgnhostPod(
+			"testpod-ovn-worker",
+			namespace,
+			"ovn-worker",
+			"/bin/bash", "-c", "/agnhost netexec --http-port 8000")
+		podWorker2 = composeAgnhostPod(
+			"testpod-ovn-worker2",
+			namespace,
+			"ovn-worker2",
+			"/bin/bash", "-c", "/agnhost netexec --http-port 8000")
+		_ = fr.PodClient().CreateSync(podWorker1)
+		_ = fr.PodClient().CreateSync(podWorker2)
+
+		By("Waiting until both pods have an IP address")
+		Eventually(func() error {
+			var err error
+			podWorker1, err = fr.ClientSet.CoreV1().Pods(fr.Namespace.Name).Get(context.TODO(), podWorker1.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if podWorker1.Status.PodIP == "" {
+				return fmt.Errorf("pod %s has no valid IP address yet", podWorker1.Name)
+			}
+			podWorker2, err = fr.ClientSet.CoreV1().Pods(fr.Namespace.Name).Get(context.TODO(), podWorker2.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if podWorker2.Status.PodIP == "" {
+				return fmt.Errorf("pod %s has no valid IP address yet", podWorker2.Name)
+			}
+			return nil
+		}).
+			WithTimeout(time.Minute).
+			WithPolling(time.Second).
+			Should(Succeed())
+
+		By("Create virtual machines")
+		vmLabels := map[string]string{}
+		if td.mode == kubevirtv1.MigrationPostCopy {
+			vmLabels = forcePostCopyMigrationPolicy.Spec.Selectors.VirtualMachineInstanceSelector
+		}
+		vms, err := composeVMs(td.numberOfVMs, vmLabels)
+		Expect(err).ToNot(HaveOccurred())
+		for _, vm := range vms {
+			vm, err = kvcli.VirtualMachine(namespace).Create(vm)
+			Expect(err).ToNot(HaveOccurred())
+		}
+		for _, vm := range vms {
+			Eventually(func() bool {
+				vm, err = kvcli.VirtualMachine(namespace).Get(vm.Name, &metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				return vm.Status.Ready
+			}).WithPolling(time.Second).WithTimeout(5 * time.Minute).Should(BeTrue())
+		}
+
+		wg.Add(int(td.numberOfVMs))
+		for _, vm := range vms {
+			go runTest(td, vm)
```

Comment:
maybe it's just me, but why do you need to run this in a routine ? 

Wouldn't it be enough to check the TCP connection from A => B ? ... at the end of the day, it's a connection. Meaning, it's actually A <=> B (there are replies).

Reply:
runTest do the following:
1.- check connectivity and network policies
2.- live migrate vm and wait for success
3.- check connectivity and network policies again 
4 .- back to bullet 2 any number of times

That's quite costly so we parallelize per VM.

Reply:
OK. One question, this number of migrations attribute ensures we put the VM back where it was ? I guess that's a side effect of having two nodes, right ? 

Maybe it makes sense to remove that attribute, and just **always** migrate the VM away from its original node, and back "home". And assure somehow the static route + policy were deleted ? (are you checking this ?)

If a deployment w/ more than 2 nodes, you could play around w/ node labels, and node selectors to assure only 2 nodes are taken into account.

Reply:
> OK. One question, this number of migrations attribute ensures we put the VM back where it was ? I guess that's a side effect of having two nodes, right ?
> 

Yep

> Maybe it makes sense to remove that attribute, and just **always** migrate the VM away from its original node, and back "home". And assure somehow the static route + policy were deleted ? (are you checking this ?)
> 

We test that at unit test under `pkg/ovn/kubevirt_test.go`

> If a deployment w/ more than 2 nodes, you could play around w/ node labels, and node selectors to assure only 2 nodes are taken into account.

Labeling a pair of nodes and using that label at the VMs should work, an alternative is to dynamically select the number of runTest with the number of nodes. Is it worth it to do it now ?

Reply:
well, but if I have 3 worker nodes it stops being "migrating back home". 

Point being - IMO - it makes sense to:
- migrate once
- migrate twice (elsewhere, then "back home")

I don't see what we get by migrating an arbitrary amount of times (or twice, but end up on a node != original node).
Anyway, I guess you have bigger fish to fry.

Reply:
Now that we have to test also multiple nodes per zone it's just migrating the number of worker zones so we ensure that last migration is going to go back to original node and that we change zones.